File: core-data/src/androidTest/java/home/oleg/placesnearme/coredata/dao/DetailedVenueWithPhotosDaoTes.java
package home.oleg.placesnearme.coredata.dao;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Collections;
import java.util.List;

import androidx.room.Room;
import androidx.test.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnit4;
import home.oleg.placesnearme.coredata.FakesStore;
import home.oleg.placesnearme.coredata.database.AppDatabase;
import home.oleg.placesnearme.coredata.model.DetailedVenueDbEntity;
import home.oleg.placesnearme.coredata.model.DetailedVenueWithPhotos;
import home.oleg.placesnearme.coredata.model.PhotoDbEntity;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;

@RunWith(AndroidJUnit4.class)
public class DetailedVenueWithPhotosDaoTes {

    private AppDatabase appDatabase;
    private DetailedVenueDao dao;

    @Before
    public void setUp() {
        appDatabase = Room
                .inMemoryDatabaseBuilder(InstrumentationRegistry.getContext(), AppDatabase.class)
                .build();
        dao = appDatabase.getDetailedVenueWithPhotosDao();
    }

    @Test
    public void shouldBeEmpty() {
        List<DetailedVenueWithPhotos> actual = dao.getAll().blockingFirst();

        assertEquals(0, actual.size());
    }

    @Test
    public void shouldInsertVenueAndPhotos() {
        DetailedVenueWithPhotos detailedVenueWithPhotos = new DetailedVenueWithPhotos();

        DetailedVenueDbEntity detailedVenue = FakesStore.getVenue("1");
        List<PhotoDbEntity> photoDbEntities = FakesStore.getListPhotos();

        detailedVenueWithPhotos.setVenue(detailedVenue);
        detailedVenueWithPhotos.setPhotos(photoDbEntities);

        dao.insert(detailedVenueWithPhotos);

        List<DetailedVenueWithPhotos> actual = dao.getAll().blockingFirst();

        assertEquals(1, actual.size());
        assertEquals(photoDbEntities.size(), actual.get(0).getPhotos().size());
    }

    @Test
    public void shouldReplaceVenueWhenInsertWithTheSameId() {
        DetailedVenueWithPhotos detailedVenueWithPhotos = new DetailedVenueWithPhotos();

        DetailedVenueDbEntity detailedVenue = FakesStore.getVenue("1");
        List<PhotoDbEntity> photoDbEntities = FakesStore.getListPhotos();

        detailedVenueWithPhotos.setPhotos(photoDbEntities);
        detailedVenueWithPhotos.setVenue(detailedVenue);

        //double inserting
        dao.insert(detailedVenueWithPhotos);
        dao.insert(detailedVenueWithPhotos);

        List<DetailedVenueWithPhotos> detailedVenueWithPhotos1 = dao.getAll().blockingFirst();

        assertEquals(1, detailedVenueWithPhotos1.size());
    }

    @Test
    public void photosShouldBeEmptyAfterReplaceVenue() {
        DetailedVenueWithPhotos detailedVenueWithPhotos = new DetailedVenueWithPhotos();

        DetailedVenueDbEntity detailedVenue = FakesStore.getVenue("1");
        List<PhotoDbEntity> photoDbEntities = FakesStore.getListPhotos();

        detailedVenueWithPhotos.setPhotos(photoDbEntities);
        detailedVenueWithPhotos.setVenue(detailedVenue);
        dao.insert(detailedVenueWithPhotos);

        //clear photos
        detailedVenueWithPhotos.setPhotos(Collections.emptyList());
        dao.insert(detailedVenueWithPhotos);

        List<PhotoDbEntity> actual = appDatabase.getDetailedVenueWithPhotosDao().getAllPhotos().blockingFirst();

        assertEquals(0, actual.size());
    }

    @Test
    public void photosShouldBeEmptyAfterDeleteVenue() {
        DetailedVenueWithPhotos detailedVenueWithPhotos = new DetailedVenueWithPhotos();

        DetailedVenueDbEntity detailedVenue = FakesStore.getVenue("1");
        List<PhotoDbEntity> photoDbEntities = FakesStore.getListPhotos();

        detailedVenueWithPhotos.setPhotos(photoDbEntities);
        detailedVenueWithPhotos.setVenue(detailedVenue);
        dao.insert(detailedVenueWithPhotos);
        dao.deleteAll();

        List<PhotoDbEntity> actual = appDatabase.getDetailedVenueWithPhotosDao().getAllPhotos().blockingFirst();

        assertEquals(0, actual.size());
    }

    @Test
    public void shouldClearAllVenues() {
        DetailedVenueWithPhotos detailedVenueWithPhotos = new DetailedVenueWithPhotos();
        DetailedVenueWithPhotos detailedVenueWithPhotos1 = new DetailedVenueWithPhotos();

        DetailedVenueDbEntity detailedVenue = FakesStore.getVenue("1");
        DetailedVenueDbEntity detailedVenue1 = FakesStore.getVenue("2");
        List<PhotoDbEntity> photoDbEntities = FakesStore.getListPhotos();

        detailedVenueWithPhotos.setPhotos(photoDbEntities);
        detailedVenueWithPhotos.setVenue(detailedVenue);

        detailedVenueWithPhotos1.setVenue(detailedVenue1);
        detailedVenueWithPhotos1.setPhotos(photoDbEntities);

        dao.insert(detailedVenueWithPhotos);
        dao.insert(detailedVenueWithPhotos1);
        dao.deleteAll();

        List<PhotoDbEntity> actualPhotos = appDatabase.getDetailedVenueWithPhotosDao().getAllPhotos().blockingFirst();
        List<DetailedVenueWithPhotos> actualVenues = dao.getAll().blockingFirst();

        assertEquals(0, actualPhotos.size());
        assertEquals(0, actualVenues.size());
    }

    @Test
    public void shouldUpdateToFavoriteVenue() {
        DetailedVenueWithPhotos detailedVenueWithPhotos = new DetailedVenueWithPhotos();

        DetailedVenueDbEntity detailedVenue = FakesStore.getVenue("1");
        detailedVenue.setFavorite(true);

        List<PhotoDbEntity> photoDbEntities = FakesStore.getListPhotos();

        detailedVenueWithPhotos.setPhotos(photoDbEntities);
        detailedVenueWithPhotos.setVenue(detailedVenue);
        dao.insert(detailedVenueWithPhotos);

        DetailedVenueDbEntity detailedVenueDbEntity = dao.getDetailedVenueById(detailedVenue.getId());
        detailedVenueDbEntity.setFavorite(false);
        dao.update(detailedVenueDbEntity);

        DetailedVenueWithPhotos actualVenue = dao.getById(detailedVenue.getId()).blockingGet();

        assertEquals(1, actualVenue.getPhotos().size());
        assertEquals(false, actualVenue.getVenue().isFavorite());
    }

    @Test
    public void shouldChangeToNotFavoriteVenue() {
        DetailedVenueWithPhotos detailedVenueWithPhotos = new DetailedVenueWithPhotos();

        DetailedVenueDbEntity detailedVenue = FakesStore.getVenue("1");
        detailedVenue.setFavorite(true);

        List<PhotoDbEntity> photoDbEntities = FakesStore.getListPhotos();

        detailedVenueWithPhotos.setPhotos(photoDbEntities);
        detailedVenueWithPhotos.setVenue(detailedVenue);
        dao.insert(detailedVenueWithPhotos);

        detailedVenueWithPhotos.getVenue().setFavorite(false);
        dao.insert(detailedVenueWithPhotos);

        DetailedVenueWithPhotos actualVenue = dao.getById("1").blockingGet();

        assertFalse(actualVenue.getVenue().isFavorite());
    }

}

File: core-data/src/androidTest/java/home/oleg/placesnearme/coredata/dao/DetailedVenueWithPhotosDaoTes.kt
package home.oleg.placesnearme.coredata.dao

import androidx.room.Room
import androidx.test.InstrumentationRegistry
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import home.oleg.placesnearme.coredata.FakesStore
import home.oleg.placesnearme.coredata.database.AppDatabase
import home.oleg.placesnearme.coredata.model.PlaceAndPhotos
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class DetailedVenueWithPhotosDaoTes {

    private lateinit var appDatabase: AppDatabase
    private lateinit var dao: PlacesDao

    @Before
    fun setUp() {
        appDatabase = Room
                .inMemoryDatabaseBuilder(ApplicationProvider.getApplicationContext(), AppDatabase::class.java)
                .build()
        dao = appDatabase.detailedVenueWithPhotosDao
    }

    @Test
    fun shouldBeEmpty() {
        val actual = dao.streamPlaces().blockingFirst()

        assertEquals(0, actual.size.toLong())
    }

    @Test
    fun shouldInsertVenueAndPhotos() {
        val detailedVenue = FakesStore.getVenue("1")
        val detailedVenueWithPhotos = PlaceAndPhotos(detailedVenue)
        detailedVenueWithPhotos.photos = FakesStore.listPhotos(detailedVenue.id)

        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)

        val actual = dao.streamPlaces().blockingFirst()

        assertEquals(1, actual.size.toLong())
        assertEquals(detailedVenueWithPhotos.photos.size.toLong(), actual[0].photos.size.toLong())
    }

    @Test
    fun shouldReplaceVenueWhenInsertWithTheSameId() {
        val detailedVenue = FakesStore.getVenue("1")
        val detailedVenueWithPhotos = PlaceAndPhotos(detailedVenue)
        detailedVenueWithPhotos.photos = FakesStore.listPhotos(detailedVenue.id)

        //double inserting
        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)
        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)

        val detailedVenueWithPhotos1 = dao.streamPlaces().blockingFirst()

        assertEquals(1, detailedVenueWithPhotos1.size.toLong())
    }

    @Test
    fun photosShouldBeEmptyAfterReplaceVenue() {
        val detailedVenue = FakesStore.getVenue("1")
        val detailedVenueWithPhotos = PlaceAndPhotos(detailedVenue)
        detailedVenueWithPhotos.photos = FakesStore.listPhotos(detailedVenue.id)

        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)

        //clear photos
        detailedVenueWithPhotos.photos = emptyList()
        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)

        val actual = appDatabase.detailedVenueWithPhotosDao.streamPhotos().blockingFirst()

        assertEquals(0, actual.size.toLong())
    }

    @Test
    fun photosShouldBeEmptyAfterDeleteVenue() {
        val detailedVenue = FakesStore.getVenue("1")
        val detailedVenueWithPhotos = PlaceAndPhotos(detailedVenue)
        detailedVenueWithPhotos.photos = FakesStore.listPhotos(detailedVenue.id)

        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)

        dao.deleteAll()

        val actual = appDatabase.detailedVenueWithPhotosDao.streamPhotos().blockingFirst()

        assertEquals(0, actual.size.toLong())
    }

    @Test
    fun shouldClearAllVenues() {

        val detailedVenue = FakesStore.getVenue("1")
        val detailedVenue1 = FakesStore.getVenue("2")

        val detailedVenueWithPhotos = PlaceAndPhotos(detailedVenue)
        detailedVenueWithPhotos.photos =  FakesStore.listPhotos(detailedVenue.id)

        val detailedVenueWithPhotos1 = PlaceAndPhotos(detailedVenue1)
        detailedVenueWithPhotos1.photos = FakesStore.listPhotos(detailedVenue1.id)

        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)
        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)
        dao.deleteAll()

        val actualPhotos = appDatabase.detailedVenueWithPhotosDao.streamPhotos().blockingFirst()
        val actualVenues = dao.streamPlaces().blockingFirst()

        assertEquals(0, actualPhotos.size.toLong())
        assertEquals(0, actualVenues.size.toLong())
    }

    @Test
    fun shouldUpdateToFavoriteVenue() {
        val detailedVenue = FakesStore.getVenue("1").copy(isFavorite = true)
        val detailedVenueWithPhotos = PlaceAndPhotos(detailedVenue)
        detailedVenueWithPhotos.photos = FakesStore.listPhotos(detailedVenue.id)

        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)

        val detailedVenueDbEntity = dao.getPlaceById(detailedVenue.id)!!.copy(isFavorite = false)

        dao.update(detailedVenueDbEntity)

        val actualVenue = dao.getById(detailedVenue.id).blockingGet()

        assertEquals(1, actualVenue.photos.size.toLong())
        assertEquals(false, actualVenue.venue.isFavorite)
    }

    @Test
    fun shouldChangeToNotFavoriteVenue() {
        val detailedVenue = FakesStore.getVenue("1").copy(isFavorite = true)
        val detailedVenueWithPhotos = PlaceAndPhotos(detailedVenue)
        detailedVenueWithPhotos.photos = FakesStore.listPhotos(detailedVenue.id)

        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)

        detailedVenueWithPhotos.venue = detailedVenueWithPhotos.venue.copy(isFavorite = false)
        dao.insertOrReplace(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)

        val (venue) = dao!!.getById("1").blockingGet()

        assertFalse(venue.isFavorite)
    }

}
